File: examples/tlv++/src/parse.cpp
// Parses a nested BER-TLV document and prints every element in document
// order. See write.cpp for building the same bytes, query.cpp for
// addressing one field directly, and validate.cpp for checking the
// document's structure without decoding it. The C, Rust and JavaScript
// "parse" examples parse the same bytes and report the same fields.
#include <array>
#include <cstddef>
#include <iomanip>
#include <iostream>

#include "tlv++/tlv.hpp"

// An FCI Template (6F) holding a DF Name (84) and an FCI Proprietary
// Template (A5) holding an Application Label (50).
static const std::array<tlv::byte, 12> document = {
    tlv::byte(0x6F), tlv::byte(0x0A), tlv::byte(0x84), tlv::byte(0x03),
    tlv::byte(0x41), tlv::byte(0x42), tlv::byte(0x43), tlv::byte(0xA5),
    tlv::byte(0x03), tlv::byte(0x50), tlv::byte(0x01), tlv::byte(0x01)};

int main() {
    size_t count = 0;
    auto   result = tlv::walk_tree(
        tlv::bytes(document.data(), document.size()), tlv_reader_format_ber, tlv_ber_is_constructed,
        TLV_WALK_MAX_DEPTH, 16, [&count](const tlv::entry& entry, size_t depth, size_t /*offset*/) {
            std::cout << std::string(depth * 2, ' ') << "tag=" << std::hex << std::uppercase
                      << static_cast<int>(entry.tag.data[0]) << " length=" << std::dec
                      << entry.value.size() << " value=" << std::hex << std::uppercase;
            for (size_t i = 0; i < entry.value.size(); ++i)
                std::cout << std::setw(2) << std::setfill('0') << static_cast<int>(entry.value[i]);
            std::cout << std::dec << "\n";
            ++count;
            return TLV_VISIT_CONTINUE;
        });
    if (!result) {
        std::cerr << "parse error: " << result.error().message << "\n";
        return 1;
    }
    // 6F, its two children (84, A5) and A5's child (50).
    return count == 4 ? 0 : 1;
}

File: bindings/lua/examples/parse.lua
--[[
Parses a nested BER-TLV document and prints every element in document
order, descending into constructed entries recursively; then shows how a
truncated buffer surfaces as a structured error instead of a partial
result. See the C, C++, Rust, JavaScript and Python "parse" examples for the
same bytes and fields.

Run with `lua examples/parse.lua` from `bindings/lua`, after building the
native module (see ../README.md or docs/development/lua.md).
]]

local opentlv = require("opentlv")

local function hex(bytes)
    local parts = {}
    for i = 1, #bytes do
        parts[i] = string.format("%02X", string.byte(bytes, i))
    end
    return table.concat(parts, " ")
end

-- Bit 6 (0x20) of the first tag byte marks a constructed (nested) entry;
-- written without Lua 5.3's `&` operator so this runs on 5.1 and LuaJIT too.
local function is_constructed(tag)
    return math.floor(string.byte(tag, 1) / 0x20) % 2 == 1
end

-- An FCI Template (6F) holding a DF Name (84) and an FCI Proprietary
-- Template (A5) holding an Application Label (50).
local DOCUMENT = string.char(0x6F, 0x0A, 0x84, 0x03, 0x41, 0x42, 0x43, 0xA5, 0x03, 0x50, 0x01, 0x01)

-- Entries borrow `data`; nesting is descended by re-reading a child entry's
-- value with a new reader.
local function print_elements(data, depth)
    depth = depth or 0
    local count = 0
    for entry in opentlv.reader(data, opentlv.formats.ber) do
        print(string.format("%stag=%s length=%d value=%s", string.rep("  ", depth),
            hex(entry.tag), entry.length, hex(entry.value)))
        count = count + 1

        if is_constructed(entry.tag) then
            count = count + print_elements(entry.value, depth + 1)
        end
    end
    return count
end

local function main()
    local count = print_elements(DOCUMENT)
    -- 6F, its two children (84, A5) and A5's child (50).
    assert(count == 4)

    -- Malformed input raises a structured error instead of a partial result.
    local truncated = DOCUMENT:sub(1, #DOCUMENT - 2)
    local ok, err = pcall(function()
        for _ in opentlv.reader(truncated, opentlv.formats.ber) do
        end
    end)
    if not ok then
        print(string.format("truncated input: %s (offset %s)", tostring(err), tostring(err.offset)))
    end
end

main()
